### PR TITLE
Added ps_process_cep to swig python interface

### PR DESCRIPTION
**Pull request by syl22-00**
This adds ps_process_cep to the Python interface in the SWIG wrapper. It is not great as it copies the values of the feature vectors, and also might not work for the JAVA bindings (not tested).

Input is given as a one-dimensional array where component i of frame j is at `((Mfcc*)SDATA)[i+j*ncep]` with `ncep` being the feature vector length.

### CODE REVIEW

**Comment by nshmyrev** [comment]
Thanks for the patch! I believe we can do typemap for this so you can just pass in float[] array from Python. So that change will require a bit more work. Do you use it this way?

**Comment by syl22-00** [comment]
Thanks @nshmyrev . I think I do use it passing arrays of float to them, This typemap thing is still very cryptic to me. Here is how I call it, it is supposed to receive feature frames in the same order as in an `.mfc` file:

```python
stream = open('my_file.mfc', 'rb')
# Skipping header
stream.read(4)
while True:
  # We should read a multiple of 4 (float size) * feature vector length 
  buffer = stream.read(1024)
  if buffer:
    decoder.process_cep(buffer, False, False)
  else:
    break
```

**Comment by mbait** [comment]
For Python accepting list of floats would imply a significant performance
impact due to packing/unpacking a list into contiguous memory blocks. For
the remaining java it's not that trivial. I can take care of the patch next
week and make typemap for Java.

On Thu, Apr 23, 2015 at 9:57 PM, Sylvain Chevalier <notifications@github.com
> wrote:

> Thanks @nshmyrev <https://github.com/nshmyrev> . I think I do use it
> passing arrays of float to them, This typemap thing is still very cryptic
> to me. Here is how I call it, it is supposed to receive feature frames in
> the same order as in an .mfc file:
>
> stream = open('my_file.mfc', 'rb')# Skipping header
> stream.read(4)while True:
>   # We should read a multiple of 4 (float size) * feature vector length
>   buffer = stream.read(1024)
>   if buffer:
>     decoder.process_cep(buffer, False, False)
>   else:
>     break
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cmusphinx/pocketsphinx/pull/15#issuecomment-95716880>.
>



-- 
Sincerely, Alexander

**Comment by syl22-00** [comment]
Is there anything I can do to help with that so that it gets somehow included? I believe it can be useful for many, for instance to build a recognition server in Python with features computed on the client side. Or just to work with features not coming from sphinxbase.

**Comment by mbait** [comment]
Sorry, I forgot completely. I'll fix tomorrow.

**Comment by syl22-00** [comment]
Thanks, let me know if I can help.

**Comment by mbait** [comment]
@syl22-00 I'm sorry, I really overloaded at work, so I don't think I can do it before the weekend. But in order to not to block you, I'd suggest to spit this patch, because the python part is already good. @nshmyrev please commit the python part then, and I'll fix the common part later.

File: swig/ps_decoder.i [hunk 1]
```diff
@@ -120,12 +120,31 @@
...
+                input[i][j] = ((Mfcc*)SDATA)[j+i*ncep];
+        return *errcode = ps_process_cep($self, input, NSAMP, no_search, full_utt);
+    }
 #else
     int
     process_raw(const int16 *SDATA, size_t NSAMP, bool no_search, bool full_utt,
                 int *errcode) {
         return *errcode = ps_process_raw($self, SDATA, NSAMP, no_search, full_utt);
     }
+    int
```
Comment (mbait) [inline swig/ps_decoder.i]: Move everything below into another patch.

File: swig/pocketsphinx.i [hunk 2]
```diff
@@ -78,6 +78,7 @@ typedef fsg_model_t FsgModel;
 typedef logmath_t LogMath;
 typedef ngram_model_t NGramModel;
 typedef ngram_model_t NGramModelSet;
+typedef mfcc_t Mfcc;
```
Comment (mbait) [inline swig/pocketsphinx.i]: Move to a  separate patch.

File: swig/ps_decoder.i [hunk 3]
```diff
@@ -120,12 +120,31 @@
         NSAMP /= sizeof(int16);
         return *errcode = ps_process_raw($self, SDATA, NSAMP, no_search, full_utt);
     }
+    int
+        process_cep(const void *SDATA, size_t NSAMP, bool no_search, bool full_utt,
+                int *errcode) {
+        int ncep = fe_get_output_size(ps_get_fe($self));
+        NSAMP /= ncep * sizeof(Mfcc);
+        Mfcc ** input = (Mfcc**) malloc(NSAMP*sizeof(Mfcc*));
```
Comment (mbait) [inline swig/ps_decoder.i]: Please use mfcc_t everywhere instead of Mfcc.

**Comment by syl22-00** [comment]
Thanks @mbait, I pushed the changes you suggested. I also added proper deallocation of memory that was missing.

**Comment by mbait** [comment]
Done. Thank you @syl22-00, please file the rest of the patch as another request. It may be empty, as it's just for me to remember.
